doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+
+- `pai.py`
+- `alisa.py`
+- `default.py`
+
+The functions `train`, `predict`, `explain` ,`run`, `solve`, `evaluate` provided by each module in `platform` should forward to `sqlflow_sumitter/xgboost`, `sqlflow_submitter/tensorflow`, etc. according to `statement.type` and `statement.estimator`. **Diagnostics are required**.
+
+The functions `train`, `predict`, `explain` ,`run`, `solve`, `evaluate `  provided by each module in `platform` should have the same signature. The signature should be defined in the detailed design doc. **Diagnostics are required**.
+
+###### Golang
+
+To execute a SQLFlow program, the golang package `pkg/executor` should eventually spawn a python process that calls `execute.py:main("execute")`, and passes the filled `Program` to stdin of the process via the standard IPC pipe.
+
+To run a step of a executing SQLFlow program, the golang package `pkg/step` should spawn a python process that calls `execute.py:main("step")`, and passes the filled `Program` (with only one `Statement`) to stdin of the process via the standard IPC pipe.
+
+To get the status of an executing SQLFlow program, the golang package `pkg/server` should eventually spawn a python process that calls `execute.py:main("fetch")` .
+
+###### MVP
+
+The python function `fetch` is **not** required in the MVP.
+
+#### Feature Derivation And Columns
+
+###### Python
+
+The function calls in the COLUMN clause should directly map to calls to python functions in the `columns` package, for example, the `COLUMN NUMERIC(x)` statement would directly map to the python function `numeric()`. We should have at least a `tf_columns.py` module under this package to implement currently supported columns such as `numeric` or `indicator`.
+
+`features.py` should have at least two function: 
+
+```python
+# Columns is the pseudo type of `map<string, Columns>` in the `Statement` message.
+def generate_analysis_steps(datasource:str, columns:Columns):
+    """
+    `generate_analysis_steps` generates a `select` statement to get necessary aggregate data for feature columns
+    """
+    pass
+  
+# ColumnsDAG is a structure to be designed in the detailed design doc.
+def generate_feature_columns(columns: ColumnsDAG):
+    """
+    `generate_feature_columns` generates python feature columns objects
+    """
+    pass
+```
+
+`generate_analysis_steps` does at least the following tasks:
+
+- Call the standard module `ast` to generate ASTs for the function call expressions in `columns`
+- Do semantic checking for every function nodes in the ASTs. **Diagnostics are required.**
+- Collect analysis requirements of the column functions from the nodes in the AST and construct analysis `select` statements
+- Construct a `ColumnsDAG` that represents the `columns` from the AST
+- Return the DAG and the analysis `select` statements to `workflow`. `workflow` should put the DAG in an Argo artifact that could be passed to the following steps. 
```

Comment:
IMO, the GO program will check the syntax and parse the `COLUMN` clause into a data structure which standards for a DAG directly.
The python module will use the data structure to do the following feature derivation, auto analysis and transform code generation.

Reply:
That's up to the syntax of the `COLUMN` clause, this part proposes we use the python function call syntax. You (and @sneaxiy ) are right if we've decided that the `COLUMN` clause are not to be compatible with the existing Python/Go function call syntax